napari/layers/utils/_slice_input.py

```diff
@@ -0,0 +1,111 @@
+from __future__ import annotations
+
+import warnings
+from dataclasses import dataclass
+from typing import List, Tuple, Union
+
+import numpy as np
+
+from napari.components.dims import reorder_after_dim_reduction
+from napari.utils.transforms import Affine
+from napari.utils.translations import trans
+
+
+@dataclass(frozen=True)
+class _SliceInput:
+    """Encapsulates the input needed for slicing a layer.
+
+    An instance of this should be associated with a layer and some of the values
+    in ``Viewer.dims`` when slicing a layer.
+    """
+
+    # The number of dimensions displayed to be displayed in the slice.
+    ndisplay: int
+    # The point in layer world coordinates that defines the slicing plane.
+    # Only the elements in the non-displayed dimensions have meaningful values.
+    point: Tuple[float, ...]
+    # The layer dimension indices in the order they are displayed.
+    # A permutation of the ``range(self.ndim)``.
+    # The last ``self.ndisplay`` dimensions are displayed in the canvas.
+    order: Tuple[int, ...]
+
+    @property
+    def ndim(self) -> int:
+        """The dimensionality of the associated layer."""
+        return len(self.order)
+
+    @property
+    def displayed(self) -> List[int]:
+        """The layer dimension indices displayed in this slice."""
+        return list(self.order[-self.ndisplay :])
+
+    @property
+    def not_displayed(self) -> List[int]:
+        """The layer dimension indices not displayed in this slice."""
+        return list(self.order[: -self.ndisplay])
+
+    def with_ndim(self, ndim: int) -> _SliceInput:
+        """Returns a new instance with the given number of layer dimensions."""
+        old_ndim = self.ndim
+        if old_ndim > ndim:
+            point = self.point[-ndim:]
+            order = reorder_after_dim_reduction(self.order[-ndim:])
+        elif old_ndim < ndim:
+            point = (0,) * (ndim - old_ndim) + self.point
+            order = tuple(range(ndim - old_ndim)) + tuple(
+                o + ndim - old_ndim for o in self.order
+            )
+        else:
+            point = self.point
+            order = self.order
+        return _SliceInput(ndisplay=self.ndisplay, point=point, order=order)
```

Comment:
am I right in saying that this becomes the 'single source of truth' for how to slice data when you have mixed dimensionality? if so I love it.

I don't love the name here - what do we think about `reduce/lower_dimensionality` as an alternative?

Reply:
No. This is really just for `Layer._update_dims`, and only does anything interesting when the `ndim` changes. We call `_update_dims` many times, though I think `ndim` will only change when the `data` setter with data of new dimensionality (which isn't something I would expect to happen much, if ever). In general, I'd love to replace lots of our usage of `_update_dims` with something like `refresh` and/or `_update_extent`.

The dimension correspondence problem is still handled by other functions/methods including `reorder_after_dim_reduction` here (which I'm not convinced is always doing something meaningful). Ideally, we would have one function to handle dimension correspondence. My sense is that we have a few things that mostly work in practice (but we know doesn't in others), but I'm still trying to understand all our usage and intended behavior there. Ultimately, having labeled dimensions would make this much simpler to understand, but we could probably do with a mostly working simple solution before then.